packages/forms/src/index.ts

```diff
@@ -204,23 +218,203 @@ export const addPatternToPage = (
+    newFieldsetId: PatternId
+  ): Blueprint => {
+    const originalFieldset = bp.patterns[originalFieldsetId] as FieldsetPattern;
+    const newFieldset = bp.patterns[newFieldsetId] as FieldsetPattern;
+    let updatedBp = { ...bp };
+
+    const idMap = new Map<PatternId, PatternId>();
+
+    for (const childId of originalFieldset.data.patterns) {
+      const childPattern = updatedBp.patterns[childId];
+      if (childPattern) {
+        const newChildPattern = copyFieldsetPattern(childPattern);
+        idMap.set(childId, newChildPattern.id);
+
+        updatedBp = {
+          ...updatedBp,
+          patterns: {
+            ...updatedBp.patterns,
+            [newChildPattern.id]: newChildPattern,
+          },
+        };
+
+        if (childPattern.type === 'fieldset') {
+          updatedBp = copyFieldsetContents(
+            updatedBp,
+            childId,
+            newChildPattern.id
+          );
+        }
+      }
+    }
+
+    newFieldset.data.patterns = originalFieldset.data.patterns.map(
+      id => idMap.get(id) || id
+    );
+
+    updatedBp = {
+      ...updatedBp,
+      patterns: {
+        ...updatedBp.patterns,
+        [newFieldsetId]: newFieldset,
+      },
+    };
+
+    return updatedBp;
+  };
+
+  let updatedBp = { ...bp };
+  let newPattern = pattern;
+
+  if (pattern.type === 'fieldset') {
+    newPattern = copyFieldsetPattern(pattern);
+  } else {
+    newPattern = copySimplePattern(pattern);
+  }
+
+  const actualParentId = findParentFieldset(bp, patternId) || parentPatternId;
+  const actualParent = updatedBp.patterns[actualParentId];
+
+  if (
+    !actualParent ||
+    !actualParent.data ||
+    !Array.isArray(actualParent.data.patterns)
+  ) {
+    throw new Error(`Invalid parent pattern with id ${actualParentId}`);
+  }
+
+  const originalIndex = actualParent.data.patterns.indexOf(patternId);
+  if (originalIndex === -1) {
+    throw new Error(
+      `Pattern with id ${patternId} not found in parent's patterns`
+    );
+  }
+
+  actualParent.data.patterns = [
+    ...actualParent.data.patterns.slice(0, originalIndex + 1),
+    newPattern.id,
+    ...actualParent.data.patterns.slice(originalIndex + 1),
+  ];
+
+  updatedBp = {
+    ...updatedBp,
+    patterns: {
+      ...updatedBp.patterns,
+      [newPattern.id]: newPattern,
+      [actualParentId]: actualParent,
+    },
+  };
+
+  if (pattern.type === 'fieldset') {
+    updatedBp = copyFieldsetContents(updatedBp, patternId, newPattern.id);
+  }
+
+  return { bp: updatedBp, pattern: newPattern };
+};
+
 export const addPatternToFieldset = (
   bp: Blueprint,
   fieldsetPatternId: PatternId,
-  pattern: Pattern
+  pattern: Pattern,
+  index?: number
 ): Blueprint => {
   const fieldsetPattern = bp.patterns[fieldsetPatternId] as FieldsetPattern;
   if (fieldsetPattern.type !== 'fieldset') {
     throw new Error('Pattern is not a page.');
   }
+
+  let updatedPagePattern: PatternId[];
+
+  if (index && index !== undefined) {
```

Comment:
Ditto, not sure if this is intended or not.

```suggestion
  if (index !== undefined) {
```

Reply:
Done.